File: solutions/python/all-your-base/1/all_your_base.py
def rebase(input_base, digits, output_base):
    if input_base < 2:
        raise ValueError("input base must be >= 2")

    if output_base < 2:
        raise ValueError("output base must be >= 2")

    if not all(0 <= d < input_base for d in digits):
        raise ValueError("all digits must satisfy 0 <= d < input base")

    in_10_base = _into_10_base(digits, input_base)
    return _into_n_base(in_10_base, output_base)

def _into_10_base(digits, input_base):
    return sum(
        d * input_base**i
        for i, d in enumerate(digits[::-1])
    )

def _into_n_base(number, output_base):
    result = []
    while number > 0:
        result.insert(0, number % output_base)
        number //= output_base

    return result or [number]

File: solutions/ruby/all-your-base/1/all_your_base.rb
class BaseConverter
  def self.convert(input_base, digits, output_base)
    raise ArgumentError if invalid_base?(input_base) || invalid_base?(output_base)
    raise ArgumentError if invalid_digits?(digits, input_base)
    
    in_10_base = new(input_base, digits, output_base).convert_into_10_base
    return in_10_base.digits.reverse if output_base == 10

    new(10, in_10_base, output_base).convert_into_given_base
  end

  def self.invalid_base?(base)
    base <= 1
  end

  def self.invalid_digits?(digits, base)
    digits.any? { |d| !(0...base).include?(d) }
  end

  public

  def convert_into_10_base
    @digits.reverse
           .each_with_index
           .reduce(0) { |sum, (digit, index)| sum + (digit * @input_base**index) }
  end

  def convert_into_given_base
    result = []

    while @digits >= @output_base do
      result << @digits % @output_base

      @digits = @digits / @output_base
    end

    result << @digits == @output_base ? 0 : @digits
    result.reverse
  end

  private

  def initialize(input_base, digits, output_base)
    @input_base = input_base
    @digits = digits
    @output_base = output_base
  end
end